c++/src/kj/compat/tls.c++

```diff
@@ -445,8 +457,53 @@ public:
   }
 
 private:
+  Promise<void> onAcceptSuccess(AuthenticatedStream&& stream) {
+    auto acceptPromise = kj::evalNow([&] {
+      // Do the SSL acceptance procedure.
+      return tls.wrapServer(kj::mv(stream));
+    });
+
+    auto sslPromise = acceptPromise.then([this](auto&& stream) -> Promise<void> {
+      // This is only attached to the success path, thus the error handler will catch if our
+      // promise fails.
+      queue.push(kj::mv(stream));
+      return kj::READY_NOW;
+    });
+    tasks.add(kj::mv(sslPromise));
+
+    // Queue up the next accept loop immediately without waiting for SSL_accept()/wrapServer().
+    return acceptLoop();
+  }
+
+  void onAcceptFailure(Exception&& e) {
+    KJ_LOG(ERROR, "TlsConnectionReceiver could not accept()", e);
+
+    // Store this exception to reject all future calls to accept() and reject any unfulfilled
+    // promises from the queue.
+    maybeInnerException = kj::mv(e);
+    queue.rejectAll(Exception(KJ_REQUIRE_NONNULL(maybeInnerException)));
+  }
+
+  Promise<void> acceptLoop() {
+    // Accept one connection and queue up the next accept on our TaskSet.
+
+    return inner->acceptAuthenticated().then(
+        [this](AuthenticatedStream&& stream) {
+      return onAcceptSuccess(kj::mv(stream));
```

Comment:
I find it a little weird that `acceptLoop()` and `onAcceptSuccess()` do mutual recursion here. It might be a little easier to see that this is a loop if `onAcceptSuccess()` returned void and here we did:

```
onAcceptSuccess(kj::mv(stream));  // will queue a handshake task to `tasks`
return acceptLoop();
```